File: frontend/src/js/api/graficos.js
/* Singra - chamadas dos gráficos
   Os números chegam prontos do backend, junto com a frase que os explica.
   Nenhuma conta financeira acontece no navegador.

   Dois deles podem voltar com disponivel = false, cartões e investimento.
   Isso não é erro: quando a pessoa não usa o recurso, a página esconde o
   bloco inteiro, em vez de mostrar um gráfico zerado.

   Requer o cliente.js. */

window.ApiGraficos = (function (global) {
    "use strict";

    var Api = global.Api;

    function categorias(ano, mes) {
        return Api.get("/api/graficos/categorias" + Api.query({ ano: ano, mes: mes }));
    }

    function cartoes(ano, mes) {
        return Api.get("/api/graficos/cartoes" + Api.query({ ano: ano, mes: mes }));
    }

    // A janela de 12 meses existe para quem tem meta de investimento anual,
    // que não fecharia dentro de 6
    function evolucao(ano, mes, meses) {
        return Api.get("/api/graficos/evolucao" +
                       Api.query({ ano: ano, mes: mes, meses: meses }));
    }

    function orcamento(ano, mes) {
        return Api.get("/api/graficos/orcamento" + Api.query({ ano: ano, mes: mes }));
    }

    function fixosVariaveis(ano, mes) {
        return Api.get("/api/graficos/fixos-variaveis" + Api.query({ ano: ano, mes: mes }));
    }

    function investimento(meses) {
        return Api.get("/api/graficos/investimento" + Api.query({ meses: meses }));
    }

    /* Busca os seis de uma vez. A página de gráficos precisa de todos, e
       pedir em paralelo evita somar uma espera atrás da outra. */
    function todos(ano, mes, meses) {
        return Promise.all([
            categorias(ano, mes),
            evolucao(ano, mes, meses),
            orcamento(ano, mes),
            fixosVariaveis(ano, mes),
            cartoes(ano, mes),
            investimento(meses),
        ]).then(function (r) {
            return {
                categorias: r[0], evolucao: r[1], orcamento: r[2],
                fixos: r[3], cartoes: r[4], investimento: r[5],
            };
        });
    }

    return {
        categorias: categorias,
        cartoes: cartoes,
        evolucao: evolucao,
        orcamento: orcamento,
        fixosVariaveis: fixosVariaveis,
        investimento: investimento,
        todos: todos,
    };
})(window);

File: backend/src/api/graficos.py
"""Singra - os dados dos gráficos.

Cada rota devolve os números já calculados e a frase que os explica. O
frontend só desenha: nenhuma conta financeira acontece no navegador, e a
mesma informação alimenta o gráfico e o texto embaixo dele.

Dois gráficos podem devolver `null` em vez de dados: o de cartões e o de
investimento. É intencional. Quando a pessoa não usa o recurso, o frontend
esconde o bloco inteiro em vez de mostrar um gráfico zerado.
"""

from flask import Blueprint, g, request

from src.api.comum import contexto_do_pedido, janela_de_meses, ok
from src.core import financas, frases
from src.core.seguranca import exigir_login

bp = Blueprint("graficos", __name__, url_prefix="/api/graficos")

# Todas as rotas daqui são protegidas, então a checagem vale para o
# blueprint inteiro em vez de repetir o decorador em cada função.
bp.before_request(exigir_login)


@bp.get("/categorias")
def por_categoria():
    ctx = contexto_do_pedido()
    tipo = request.args.get("tipo", "gasto")
    dados = financas.grafico_categorias(
        g.usuario, ctx, tipo if tipo in ("gasto", "ganho") else "gasto"
    )
    return ok(**dados, frase=frases.categorias(dados))


@bp.get("/cartoes")
def por_cartao():
    """Devolve disponivel=False quando não há cartão cadastrado."""
    ctx = contexto_do_pedido()
    dados = financas.grafico_cartoes(g.usuario, ctx)
    if dados is None:
        return ok(disponivel=False, itens=[], total=0)
    return ok(disponivel=True, **dados, frase=frases.cartoes(dados))


@bp.get("/evolucao")
def evolucao():
    ctx = contexto_do_pedido()
    dados = financas.grafico_evolucao(g.usuario, ctx, janela_de_meses())
    return ok(**dados, frase=frases.evolucao(dados), frase_saldo=frases.saldo(dados))


@bp.get("/orcamento")
def orcamento():
    ctx = contexto_do_pedido()
    dados = financas.grafico_orcamento(g.usuario, ctx)
    return ok(**dados, frase=frases.orcamento(dados))


@bp.get("/fixos-variaveis")
def fixos_variaveis():
    ctx = contexto_do_pedido()
    dados = financas.grafico_fixos_variaveis(g.usuario, ctx)
    return ok(**dados, frase=frases.fixos_variaveis(dados))


@bp.get("/investimento")
def investimento():
    """Evolução do guardado contra a meta, período a período da cadência."""
    dados = financas.grafico_investimento(g.usuario, janela_de_meses())
    if dados is None:
        return ok(disponivel=False, itens=[])
    return ok(disponivel=True, **dados, frase=frases.investimento(dados))
